test/test_compile_package.py

```diff
@@ -0,0 +1,374 @@
+
+        args = torch.randn(3), torch.randn(3)
+        args1 = torch.randn(4), torch.randn(4)
+        f(*args)
+        expected1 = f(*args1)
+        f = torch.compile(f, fullgraph=True, dynamic=False, package=self.path())
+        f(*args)
+        f.save_package()
+        with self.assertRaisesRegex(RuntimeError, "Detected recompile"):
+            self._test_load(f, args1, expected1, self.path())
+
+    @requires_cuda
+    @parametrize("device", ["cpu", "cuda"])
+    def test_basic_function_different_dir(self, device):
+        torch.set_default_device(device)
+
+        def f(a, b):
+            for i in range(30):
+                a = a + b * i
+            return a
+
+        args = torch.randn(3), torch.randn(3)
+        expected = f(*args)
+        f = torch.compile(f, fullgraph=True, dynamic=False, package=self.path())
+        f(*args)
+        f.save_package()
+        new_prefix = os.path.join(cache_dir(), "new")
+        os.makedirs(new_prefix, exist_ok=False)
+        shutil.move(self.path(), new_prefix)
+
+        with self.assertRaises(RuntimeError):
+            f.load_package()
+
+        new_path = os.path.join(new_prefix, os.path.split(self.path())[-1])
+        f = torch.compile(f, fullgraph=True, dynamic=False, package=new_path)
+        self._test_load(f, args, expected, new_path)
+
+    def test_basic_function_tensor_meta_mismatch(self):
+        def f(a, b):
+            for i in range(30):
+                a = a + b * i
+            return a
+
+        args = torch.randn(3), torch.randn(3)
+        expected = f(*args)
+        f = torch.compile(f, fullgraph=True, dynamic=False, package=self.path())
+        f(*args)
+        self._test_save(f, args, expected, self.path(), 0)
+        with self.assertRaisesRegex(RuntimeError, "Detected recompile"):
+            self._test_load(f, (torch.randn(3), torch.randn(4)), None, self.path())
+        with self.assertRaisesRegex(RuntimeError, "Detected recompile"):
+            self._test_load(
+                f, (torch.zeros(3), torch.zeros(3, dtype=torch.int)), None, self.path()
+            )
+
+    @unittest.expectedFailure("failing with AOTI right now")
+    def test_basic_function_dynamic_shape(self):
+        def f(a, b):
+            for i in range(30):
+                a = a + b * i
+            return a
+
+        args = torch.randn(3), torch.randn(3)
+        args1 = torch.randn(4, 2), torch.randn(4, 2)
+        args2 = torch.randn(4), torch.randn(4)
+        args3 = torch.randn(5), torch.randn(5)
+        expected = f(*args)
+        expected1 = f(*args1)
+        expected2 = f(*args2)
+        expected3 = f(*args3)
+        f = torch.compile(f, fullgraph=True, package=self.path())
+        f(*args)
+        f(*args1)
+        f(*args2)
+
+        self._test_save(f, args3, expected3, self.path(), 2)
+        self._test_load(f, args3, expected3, self.path())
+
+        # self._test_save(f, (torch.randn(4, 2), torch.randn(4, 3)), None, self.path(), None)
+
```

Comment:
Can we add a test for trying to load a saved package generated for function `foo` on function `bar`.
1) If the name/signature don't match, we should throw an error on load
2) We should also recursively hash the source code and assert the source codes match.  (With a flag to disable this check)